File: Demos/ObjCDemo/Sources/DemoTheme.m
#import "DemoTheme.h"

static UIColor *PYKHexColor(uint32_t hex) {
    return [UIColor colorWithRed:((hex >> 16) & 0xFF) / 255.0
                           green:((hex >> 8) & 0xFF) / 255.0
                            blue:(hex & 0xFF) / 255.0
                           alpha:1.0];
}

@implementation DemoTheme

+ (UIColor *)canvas { return PYKHexColor(0x0A0C10); }
+ (UIColor *)card { return PYKHexColor(0x14171F); }
+ (UIColor *)cardElevated { return PYKHexColor(0x1B1F29); }
+ (UIColor *)cardBorder { return PYKHexColor(0x252A36); }
+ (UIColor *)field { return PYKHexColor(0x0E1117); }
+ (UIColor *)textPrimary { return PYKHexColor(0xF3F5FA); }
+ (UIColor *)textSecondary { return PYKHexColor(0x8B93A7); }
+ (UIColor *)textTertiary { return PYKHexColor(0x5C6378); }
+ (UIColor *)accent { return PYKHexColor(0xFF5C39); }
+ (UIColor *)wechat { return PYKHexColor(0x07C160); }
+ (UIColor *)alipay { return PYKHexColor(0x1677FF); }
+ (UIColor *)success { return PYKHexColor(0x2BD576); }
+ (UIColor *)failure { return PYKHexColor(0xFF5C5C); }
+ (UIColor *)cancelled { return PYKHexColor(0xFFB23E); }

+ (CGFloat)cardRadius { return 18.0; }
+ (CGFloat)controlRadius { return 14.0; }
+ (CGFloat)pagePadding { return 18.0; }

+ (UIFont *)roundedFontOfSize:(CGFloat)size weight:(UIFontWeight)weight {
    UIFont *base = [UIFont systemFontOfSize:size weight:weight];
    UIFontDescriptor *descriptor = [base.fontDescriptor fontDescriptorWithDesign:UIFontDescriptorSystemDesignRounded];
    if (descriptor == nil) {
        return base;
    }
    return [UIFont fontWithDescriptor:descriptor size:size];
}

+ (UIFont *)textFontOfSize:(CGFloat)size weight:(UIFontWeight)weight {
    return [UIFont systemFontOfSize:size weight:weight];
}

+ (UIView *)cardView {
    UIView *view = [[UIView alloc] init];
    view.backgroundColor = DemoTheme.card;
    view.layer.cornerRadius = DemoTheme.cardRadius;
    view.layer.cornerCurve = kCACornerCurveContinuous;
    view.layer.borderWidth = 1.0;
    view.layer.borderColor = DemoTheme.cardBorder.CGColor;
    view.translatesAutoresizingMaskIntoConstraints = NO;
    return view;
}

+ (UIButton *)primaryButtonWithTitle:(NSString *)title {
    UIButton *button = [UIButton buttonWithType:UIButtonTypeSystem];
    UIButtonConfiguration *configuration = [UIButtonConfiguration filledButtonConfiguration];
    configuration.baseBackgroundColor = DemoTheme.accent;
    configuration.baseForegroundColor = UIColor.whiteColor;
    configuration.cornerStyle = UIButtonConfigurationCornerStyleLarge;
    configuration.contentInsets = NSDirectionalEdgeInsetsMake(17.0, 16.0, 17.0, 16.0);
    NSAttributedString *attributed = [[NSAttributedString alloc]
        initWithString:title
            attributes:@{ NSFontAttributeName: [DemoTheme roundedFontOfSize:17.0 weight:UIFontWeightSemibold] }];
    configuration.attributedTitle = attributed;
    button.configuration = configuration;
    button.layer.cornerCurve = kCACornerCurveContinuous;
    button.configurationUpdateHandler = ^(UIButton *updateButton) {
        updateButton.alpha = updateButton.highlighted ? 0.85 : 1.0;
        updateButton.transform = updateButton.highlighted
            ? CGAffineTransformMakeScale(0.98, 0.98)
            : CGAffineTransformIdentity;
    };
    return button;
}

@end

File: Demos/SwiftDemo/Sources/DemoTheme.swift
import UIKit

/// Centralized visual language for the demo. The demo locks to a custom dark
/// "premium fintech" palette so the look is intentional rather than inheriting
/// system grouped-table chrome.
enum DemoTheme {
    // MARK: Palette

    static let canvas = UIColor(hex: 0x0A0C10)
    static let card = UIColor(hex: 0x14171F)
    static let cardElevated = UIColor(hex: 0x1B1F29)
    static let cardBorder = UIColor(hex: 0x252A36)
    static let field = UIColor(hex: 0x0E1117)

    static let textPrimary = UIColor(hex: 0xF3F5FA)
    static let textSecondary = UIColor(hex: 0x8B93A7)
    static let textTertiary = UIColor(hex: 0x5C6378)

    /// Brand accent / primary CTA. Distinct from both channel colors so the
    /// "去支付" action never reads as a specific channel.
    static let accent = UIColor(hex: 0xFF5C39)
    static let accentInk = UIColor.white

    static let wechat = UIColor(hex: 0x07C160)
    static let alipay = UIColor(hex: 0x1677FF)

    static let success = UIColor(hex: 0x2BD576)
    static let failure = UIColor(hex: 0xFF5C5C)
    static let cancelled = UIColor(hex: 0xFFB23E)

    // MARK: Metrics

    static let cardRadius: CGFloat = 18
    static let controlRadius: CGFloat = 14
    static let pagePadding: CGFloat = 18

    // MARK: Fonts

    /// Rounded SF for numerals and the wordmark — friendlier, more premium than
    /// the default cut, and the most distinctive typographic move available
    /// without bundling a custom face.
    static func rounded(_ size: CGFloat, _ weight: UIFont.Weight) -> UIFont {
        let base = UIFont.systemFont(ofSize: size, weight: weight)
        guard let descriptor = base.fontDescriptor.withDesign(.rounded) else {
            return base
        }
        return UIFont(descriptor: descriptor, size: size)
    }

    static func text(_ size: CGFloat, _ weight: UIFont.Weight) -> UIFont {
        UIFont.systemFont(ofSize: size, weight: weight)
    }

    // MARK: Builders

    static func card(cornerRadius: CGFloat = cardRadius) -> UIView {
        let view = UIView()
        view.backgroundColor = card
        view.layer.cornerRadius = cornerRadius
        view.layer.cornerCurve = .continuous
        view.layer.borderWidth = 1
        view.layer.borderColor = cardBorder.cgColor
        view.translatesAutoresizingMaskIntoConstraints = false
        return view
    }

    static func primaryButton(title: String) -> UIButton {
        let button = UIButton(type: .system)
        var configuration = UIButton.Configuration.filled()
        configuration.baseBackgroundColor = accent
        configuration.baseForegroundColor = accentInk
        configuration.cornerStyle = .large
        configuration.contentInsets = NSDirectionalEdgeInsets(top: 17, leading: 16, bottom: 17, trailing: 16)
        configuration.attributedTitle = AttributedString(
            title,
            attributes: AttributeContainer([.font: rounded(17, .semibold)])
        )
        button.configuration = configuration
        button.layer.cornerCurve = .continuous
        // Subtle press feedback.
        button.configurationUpdateHandler = { button in
            button.alpha = button.isHighlighted ? 0.85 : 1
            button.transform = button.isHighlighted
                ? CGAffineTransform(scaleX: 0.98, y: 0.98)
                : .identity
        }
        return button
    }
}

extension UIColor {
    convenience init(hex: UInt32, alpha: CGFloat = 1) {
        self.init(
            red: CGFloat((hex >> 16) & 0xFF) / 255,
            green: CGFloat((hex >> 8) & 0xFF) / 255,
            blue: CGFloat(hex & 0xFF) / 255,
            alpha: alpha
        )
    }
}
